### PR TITLE
gh-188: add docstrings to all functions and tidy docs

### LINKED ISSUE #188: Add docstrings to all functions
Currently, most have them but those that are lacking do stick out in the documentation - which automatically pulls from them

### PR DESCRIPTION
**Pull request by paddyroddy**
Closes #188, #369, #375. Have set #358 as upstream as it eases the process.

Adds docstrings to all functions, namely a description for all parameters in the `glass` package. I don't know the code base well, so would be good for these to be checked that they make sense.

I'm also not really familiar with RST so worth checking if any of that can be improved.

### CODE REVIEW

File: glass/shells.py [hunk 1]
```diff
@@ -413,6 +481,7 @@ def partition(
     approximate solution of
 
     .. math::
+
         \begin{pmatrix}
```
Comment (paddyroddy) [inline glass/shells.py:485]: This was previously breaking the maths

File: docs/conf.py [hunk 2]
```diff
@@ -62,8 +64,8 @@
 # This config value contains the locations and names of other projects that
 # should be linked to in this documentation.
 intersphinx_mapping = {
+    "numpy": ("https://numpy.org/doc/stable", None),
     "python": ("https://docs.python.org/3", None),
-    "numpy": ("https://numpy.org/doc/stable/", None),
 }
```
Comment (paddyroddy) [inline docs/conf.py:69]: See #350

File: glass/lensing.py [hunk 3]
```diff
@@ -61,27 +61,29 @@ def from_convergence(  # noqa: PLR0913
...
+    discretized
         Correct the pixel window function in output maps.
 
+    Returns
+    -------
+        The maps of:
+
+        * deflection potential if ``potential`` is true.
+        * potential (complex) if ``deflection`` is true.
+        * shear (complex) if ``shear`` is true.
```
Comment (Saransh-cpp) [inline glass/lensing.py:85]: Maybe requires better indentation?

image

```suggestion
        The maps of:

            * deflection potential if ``potential`` is true.
            * potential (complex) if ``deflection`` is true.
            * shear (complex) if ``shear`` is true.
```

File: glass/lensing.py [hunk 4]
```diff
@@ -61,27 +61,29 @@ def from_convergence(  # noqa: PLR0913
...
+
     Notes
```
Comment (Saransh-cpp) [inline glass/lensing.py:87]: Nitpicking, but "Notes" and "References" in the rendered docs do not get a `:` by default -

image

Should we add it in the docstring, or just leave it as it is?

**Review by Saransh-cpp — COMMENTED** [review]
Amazing, thanks @paddyroddy! Quite a lot of efforts in this 🙂

**Review by Saransh-cpp — APPROVED** [review]

File: glass/lensing.py (hunk 3 above)
Comment (paddyroddy) [reply glass/lensing.py]: Good idea

File: glass/lensing.py (hunk 4 above)
Comment (paddyroddy) [reply glass/lensing.py:87]: I had spotted this. I'd be inclined to follow the NumpyDoc standard https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard

**Comment by paddyroddy** [comment]
@ntessore can you have a look?

File: glass/shells.py [hunk 5]
```diff
@@ -538,6 +631,21 @@ def partition_nnls(
     Uses the ``nnls()`` algorithm from ``scipy.optimize`` and thus
     requires SciPy.
```
Comment (ntessore) [inline glass/shells.py:633]: This isn't actually true, which took a lot of effort :)

**Review by ntessore — COMMENTED** [review]
Excellent, this must have been hard work! A few trivial things noted below

File: glass/shells.py (hunk 5 above)
Comment (paddyroddy) [reply glass/shells.py]: Nice, have removed that sentence

File: glass/points.py [hunk 6]
```diff
@@ -148,37 +178,38 @@ def positions_from_delta(  # noqa: PLR0912, PLR0913, PLR0915
...
-    lon:
-        Columns of longitudes for the sampled points.
-    lat:
-        Columns of latitudes for the sampled points.
-    count:
-        The number of sampled points  If multiple populations are sampled, an
-        array of counts in the shape of the extra dimensions is returned.
+        The longitudes, latitudes and the number of sampled points. If
+        multiple populations are sampled, an array of counts in the shape
+        of the extra dimensions is returned.
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply glass/points.py:207]: I feared you might say this. Will see what I can do. Seems to be some sphinx/autodoc oddity.

[the remark this answers is not in the document]
Comment (ntessore) [reply glass/points.py:207]: Not sure if this works for the yields section, but worth a try maybe:

```py
# conf.py
napoleon_custom_sections = [("Yields", "params_style")]
```

[the remark this answers is not in the document]
Comment (paddyroddy) [reply glass/points.py]: I found a way to do it without fiddling with `napolean`. It seems `Returns` and `Yields` operate slightly differently.

File: glass/points.py [hunk 7]
```diff
@@ -300,20 +331,15 @@ def uniform_positions(
...
     ------
-    lon:
-        Columns of longitudes for the sampled points.
-    lat:
-        Columns of latitudes for the sampled points.
-    count:
-        The number of sampled points. For array inputs, an array of
-        counts with the same shape is returned.
+        The longitudes, latitudes and the number of sampled points. For array
+        inputs, an array of counts with the same shape is returned.
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply glass/points.py]: Fixed https://github.com/glass-dev/glass/pull/381#discussion_r1840492686

**Comment by paddyroddy** [comment]
@ntessore this is ready again

**Comment by ntessore** [comment]
The yields sections are rendering the text as types now.  The original formatting actually produces the desired output:

```
    Yields
    ------
    lon: ndarray[Any, dtype[float64]]
        Columns of longitudes for the sampled points.
    lat: ndarray[Any, dtype[float64]]
        Columns of latitudes for the sampled points.
    count: int | ndarray[Any, dtype[int64]]
        The number of sampled points  If multiple populations are sampled, an
        array of counts in the shape of the extra dimensions is returned.
```

You could leave out the types here if you are concerned about duplicating the information, but I think there's little harm in these isolated cases.

**Comment by paddyroddy** [comment]
What I wanted to do is what we've done for `from_convergence`, but I've tried every variant, and it doesn't work. This is what I mean when I say `Yields` act differently.

image

**Comment by ntessore** [comment]
Or even, using

```py
# conf.py
napoleon_custom_sections = [
    ("Returns", "params_style"),
    ("Yields", "params_style"),
]
```

having it apply the correct bold font to the variable names:

image

This also seems to change the order to Parameters - Yields - Returns - Raises - Return Type which keeps the most relevant information together.

**Comment by paddyroddy** [comment]
> But even if you don't want to repeat the types, what's wrong with this?

I might be imagining it, but I thought we'd discussed that syntax previously. Which is why I wasn't using it.

It looks like it does work. I guess the main drawback is that it implies a name for the output, which is in general up to the user. It would also make sense to do this throughout for the codebase for all `Returns`/`Yields`. Is that what you would like?

**Comment by ntessore** [comment]
> I might be imagining it, but I thought we'd discussed that syntax previously. Which is why I wasn't using it.

As I remember it, we discussed this when we tried to get the return type into the returns section. Since that wasn't working well, I thought we decided to keep the return type section separate, I think there's nothing in the way of having multiple yields and returns?

**Comment by paddyroddy** [comment]
> As I remember it, we discussed this when we tried to get the return type into the returns section. Since that wasn't working well, I thought we decided to keep the return type section separate, I think there's nothing in the way of having multiple yields and returns?

Yeah it must have been then. I will go through and change all `Returns`/`Yields` now.

**Comment by paddyroddy** [comment]
> Yeah it must have been then. I will go through and change all `Returns`/`Yields` now.

Done. It looks like it was just that one `Returns` and those two `Yields`.

File: glass/lensing.py [hunk 8]
```diff
@@ -78,11 +78,12 @@ def from_convergence(  # noqa: PLR0913
...
-
-            * potential (complex) if ``deflection`` is true.
-            * deflection potential if ``potential`` is true.
-            * shear (complex) if ``shear`` is true.
+    alpha
+        Map of the deflection (complex). Only returned if ``deflection`` if true.
+    psi
+        Map of the deflection potential. Only returned if ``potential`` is true.
+    gamma
+        Map of the shear (complex). Only returned if ``shear`` is true.
```
Comment (ntessore) [inline glass/lensing.py:86]: My apologies, it seems like I was confusing myself before. The order is `psi, alpha, gamma` like the order of the derivatives.

```suggestion
    psi
        Map of the lensing (or deflection) potential. Only returned if ``potential`` is true.
    alpha
        Map of the deflection (complex). Only returned if ``deflection`` if true.
    gamma
        Map of the shear (complex). Only returned if ``shear`` is true.
```

**Review by ntessore — APPROVED** [review]
Fantastic effort, particularly with the many changes in specs! There is a remaining issue with single-line yields but it's so minuscule that I want to tackle it separately. Towards a new release!

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @paddyroddy! Looks good and just needs fixing the minor pre-commit.ci failure.

**Comment by paddyroddy** [comment]
pre-commit.ci autofix

**Comment by paddyroddy** [comment]
> Fantastic effort, particularly with the many changes in specs! There is a remaining issue with single-line yields but it's so minuscule that I want to tackle it separately. Towards a new release!

Great, could you make an issue for this?

**Comment by paddyroddy** [comment]
> Great, could you make an issue for this?

https://github.com/glass-dev/glass/issues/418

**Pull request merged**